File: C++/2.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode *l3 = NULL; // 结果的头指针
        ListNode **p = &l3; // 指向（指向链表最后一个元素的下一个元素的指针）的指针，便于构建链表
        int carry = 0; // 进位
        while(l1!=NULL || l2!=NULL)
        {
            // 如果两个链表都到达尾部则跳出循环
            if(l1==NULL && l2==NULL) break;

            // 计算本位（和上一位的进位）的和
            int s = carry;
            if(l1) s += l1->val, l1 = l1->next;
            if(l2) s += l2->val, l2 = l2->next;

            // 为结果建立新的结点，并计算进位
            *p = new ListNode(s%10);
            carry = s/10;

            // 移动指针到下一个位置
            p = &((*p)->next);
        }
        // 如果还有进位，则还需再建立一个结点
        if(carry != 0)
            *p = new ListNode(carry);
        return l3;
    }
};

File: Python/2.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        head = ListNode(0)
        val = 0
        p = head
        while True:
            if l1 is None and l2 is None: break
            if l1:
                val += l1.val
                l1 = l1.next
            if l2:
                val += l2.val
                l2 = l2.next
            p.next = ListNode(val % 10)
            val /= 10
            p = p.next
        if val > 0:
            p.next = ListNode(val)
        return head.next

'''
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param {ListNode} l1
    # @param {ListNode} l2
    # @return {ListNode}
    def addTwoNumbers(self, l1, l2):
        def conver(l):
            li = []
            while l:
                li.append(l.val)
                l=l.next
            return int(''.join(map(str,li[::-1])))

        li1 = conver(l1)
        li2 = conver(l2)
        res = str(li1+li2)[::-1]
        head = ListNode(res[0])
        p=head
        if len(res) ==1:
            return head
        for i in range(len(res)-1):
                q = ListNode(res[i+1])
                p.next=q
                p=q

        return head
'''
